File: uniffi_bindgen/src/bindings/swift/templates/CallbackInterfaceImpl.swift
{%- let trait_impl=format!("UniffiCallbackInterface{}", name) %}

// Put the implementation in a struct so we don't pollute the top-level namespace
fileprivate struct {{ trait_impl }} {

    // Create the VTable using a series of closures.
    // Swift automatically converts these into C callback functions.
    //
    // This creates 1-element array, since this seems to be the only way to construct a const
    // pointer that we can pass to the Rust code.
    static let vtable: [{{ vtable|ffi_type_name }}] = [{{ vtable|ffi_type_name }}(
        uniffiFree: { (uniffiHandle: UInt64) -> () in
            do {
                try {{ ffi_converter_name }}.handleMap.remove(handle: uniffiHandle)
            } catch {
                print("Uniffi callback interface {{ name }}: handle missing in uniffiFree")
            }
        },
        uniffiClone: { (uniffiHandle: UInt64) -> UInt64 in
            do {
                return try {{ ffi_converter_name }}.handleMap.clone(handle: uniffiHandle)
            } catch {
                fatalError("Uniffi callback interface {{ name }}: handle missing in uniffiClone")
            }
        },
        {%- for (ffi_callback, meth) in vtable_methods %}
        {{ meth.name()|fn_name }}: { (
            {%- for arg in ffi_callback.arguments() %}
            {{ arg.name()|var_name }}: {{ arg.type_().borrow()|ffi_type_name }}{% if !loop.last || ffi_callback.has_rust_call_status_arg() %},{% endif %}
            {%- endfor -%}
            {%- if ffi_callback.has_rust_call_status_arg() %}
            uniffiCallStatus: UnsafeMutablePointer<RustCallStatus>
            {%- endif %}
        ) in
            let makeCall = {
                () {% if meth.is_async() %}async {% endif %}throws -> {% match meth.return_type() %}{% when Some(t) %}{{ t|type_name }}{% when None %}(){% endmatch %} in
                guard let uniffiObj = try? {{ ffi_converter_name }}.handleMap.get(handle: uniffiHandle) else {
                    throw UniffiInternalError.unexpectedStaleHandle
                }
                return {% if meth.throws() %}try {% endif %}{% if meth.is_async() %}await {% endif %}uniffiObj.{{ meth.name()|fn_name }}(
                    {%- for arg in meth.arguments() %}
                    {% if !config.omit_argument_labels() %} {{ arg.name()|arg_name }}: {% endif %}try {{ arg|lift_fn }}({{ arg.name()|var_name }}){% if !loop.last %},{% endif %}
                    {%- endfor %}
                )
            }
            {%- if !meth.is_async() %}

            {% match meth.return_type() %}
            {%- when Some(t) %}
            let writeReturn = { uniffiOutReturn.pointee = {{ t|lower_fn }}($0) }
            {%- when None %}
            let writeReturn = { () }
            {%- endmatch %}

            {%- match meth.throws_type() %}
            {%- when None %}
            uniffiTraitInterfaceCall(
                callStatus: uniffiCallStatus,
                makeCall: makeCall,
                writeReturn: writeReturn
            )
            {%- when Some(error_type) %}
            uniffiTraitInterfaceCallWithError(
                callStatus: uniffiCallStatus,
                makeCall: makeCall,
                writeReturn: writeReturn,
                lowerError: {{ error_type|lower_fn }}
            )
            {%- endmatch %}
            {%- else %}

            let uniffiHandleSuccess = { (returnValue: {{ meth.return_type()|return_type_name }}) in
                uniffiFutureCallback(
                    uniffiCallbackData,
                    {{ meth.foreign_future_ffi_result_struct().name()|ffi_struct_name }}(
                        {%- if let Some(return_type) = meth.return_type() %}
                        returnValue: {{ return_type|lower_fn }}(returnValue),
                        {%- endif %}
                        callStatus: RustCallStatus()
                    )
                )
            }
            let uniffiHandleError = { (statusCode, errorBuf) in
                uniffiFutureCallback(
                    uniffiCallbackData,
                    {{ meth.foreign_future_ffi_result_struct().name()|ffi_struct_name }}(
                        {%- if let Some(return_type) = meth.return_type() %}
                        returnValue: {{ meth.return_type().map(FfiType::from)|ffi_default_value }},
                        {%- endif %}
                        callStatus: RustCallStatus(code: statusCode, errorBuf: errorBuf)
                    )
                )
            }

            {%- match meth.throws_type() %}
            {%- when None %}
            uniffiTraitInterfaceCallAsync(
                makeCall: makeCall,
                handleSuccess: uniffiHandleSuccess,
                handleError: uniffiHandleError,
                droppedCallback: uniffiOutDroppedCallback
            )
            {%- when Some(error_type) %}
            uniffiTraitInterfaceCallAsyncWithError(
                makeCall: makeCall,
                handleSuccess: uniffiHandleSuccess,
                handleError: uniffiHandleError,
                lowerError: {{ error_type|lower_fn }},
                droppedCallback: uniffiOutDroppedCallback
            )
            {%- endmatch %}
            {%- endif %}
        }{% if !loop.last %},{% endif %}
        {%- endfor %}
    )]
}

private func {{ callback_init }}() {
    {{ ffi_init_callback.name() }}({{ trait_impl }}.vtable)
}

File: uniffi_bindgen/src/bindings/python/templates/CallbackInterfaceImpl.py
# Put all the bits inside a class to keep the top-level namespace clean
class {{ trait_impl }}:
    # For each method, generate a callback function to pass to Rust
    {%- for meth in vtable.methods %}
    {%- let callable = meth.callable %}

    @{{ meth.ffi_type.type_name }}
    def {{ callable.name }}(
            uniffi_handle,
            {%- for arg in callable.arguments %}
            {{ arg.name }},
            {%- endfor -%}
            {%- if !callable.is_async %}
            uniffi_out_return,
            uniffi_call_status_ptr,
            {%- else %}
            uniffi_future_callback,
            uniffi_callback_data,
            uniffi_out_return,
            {%- endif %}
        ):
        uniffi_obj = {{ ffi_converter_name }}._handle_map.get(uniffi_handle)
        def make_call():
            args = ({% for arg in callable.arguments %}{{ arg.ty.ffi_converter_name }}.lift({{ arg.name }}), {% endfor %})
            method = uniffi_obj.{{ callable.name }}
            return method(*args)

        {%- match callable.async_data %}
        {%- when None %}
        {%- match callable.return_type.ty %}
        {%- when Some(return_type) %}
        def write_return_value(v):
            uniffi_out_return[0] = {{ return_type.ffi_converter_name }}.lower(v)
        {%- when None %}
        write_return_value = lambda v: None
        {%- endmatch %}

        {%- match callable.throws_type.ty %}
        {%- when None %}
        _uniffi_trait_interface_call(
                uniffi_call_status_ptr.contents,
                make_call,
                write_return_value,
        )
        {%- when Some(error) %}
        _uniffi_trait_interface_call_with_error(
                uniffi_call_status_ptr.contents,
                make_call,
                write_return_value,
                {{ error.type_name }},
                {{ error.ffi_converter_name }}.lower,
        )
        {%- endmatch %}
        {%- when Some(async_data) %}
        def handle_success(return_value):
            uniffi_future_callback(
                uniffi_callback_data,
                {{ async_data.ffi_foreign_future_result.0 }}(
                    {%- if let Some(return_type) = callable.return_type.ty %}
                    {{ return_type.ffi_converter_name }}.lower(return_value),
                    {%- endif %}
                    _UniffiRustCallStatus.default()
                )
            )

        def handle_error(status_code, rust_buffer):
            uniffi_future_callback(
                uniffi_callback_data,
                {{ async_data.ffi_foreign_future_result.0 }}(
                    {%- match callable.return_type.ty %}
                    {%- when Some(return_type) %}
                    {{ meth.ffi_default_value }},
                    {%- when None %}
                    {%- endmatch %}
                    _UniffiRustCallStatus(status_code, rust_buffer),
                )
            )

        {%- match callable.throws_type.ty %}
        {%- when None %}
        uniffi_out_return[0] = _uniffi_trait_interface_call_async(make_call, handle_success, handle_error)
        {%- when Some(error) %}
        uniffi_out_return[0] = _uniffi_trait_interface_call_async_with_error(
            make_call,
            handle_success,
            handle_error,
            {{ error.type_name }},
            {{ error.ffi_converter_name }}.lower,
        )
        {%- endmatch %}
        {%- endmatch %}
    {%- endfor %}

    @{{ vtable.free_fn_type.0 }}
    def _uniffi_free(uniffi_handle):
        {{ ffi_converter_name }}._handle_map.remove(uniffi_handle)

    # Generate the FFI VTable.  This has a field for each callback interface method.
    _uniffi_vtable = {{ vtable.struct_type.type_name }}(
        {%- for meth in vtable.methods %}
        {{ meth.callable.name }},
        {%- endfor %}
        _uniffi_free
    )
    # Send Rust a pointer to the VTable.  Note: this means we need to keep the struct alive forever,
    # or else bad things will happen when Rust tries to access it.
    _UniffiLib.{{ vtable.init_fn.0 }}(ctypes.byref(_uniffi_vtable))
